src/containers/App/App.js

```diff
@@ -93,39 +94,91 @@ if (process.env.I18N_PSEUDO) {
   });
 }
 
+const initialState = {
+  loadingConfigError: null,
+  loadingConfig: true
+};
+
 export /* istanbul ignore next */ class App extends Component {
+  state = initialState;
+
   componentDidMount() {
-    this.fetchData();
+    this.fetchConfig();
   }
 
-  componentDidUpdate(prevProps) {
+  componentDidUpdate(prevProps, prevState) {
     const { webSocketConnected } = this.props;
     const { webSocketConnected: prevWebSocketConnected } = prevProps;
+
     if (webSocketConnected && prevWebSocketConnected === false) {
-      this.fetchData();
+      this.fetchConfig();
+    }
+
+    const { loadingConfig } = this.state;
+    const { loadingConfig: prevLoadingConfig } = prevState;
+
+    if (prevLoadingConfig && loadingConfig === false) {
+      if (this.props.tenantNamespace) {
+        this.props.selectNamespace(this.props.tenantNamespace);
+      }
+
+      if (!this.props.tenantNamespace) {
+        this.props.fetchNamespaces();
+      }
+
+      this.props.fetchExtensions({
+        namespace: this.props.tenantNamespace
+      });
     }
   }
 
   componentWillUnmount() {
     this.props.onUnload();
   }
 
-  fetchData() {
-    this.props.fetchExtensions();
-    this.props.fetchInstallProperties();
-    this.props.fetchNamespaces();
+  async fetchConfig() {
+    this.setState({ loadingConfig: true });
+    try {
+      this.props.fetchInstallProperties();
+      this.setState({ loadingConfig: false });
+    } catch (error) {
+      console.error(error); // eslint-disable-line no-console
+      this.setState({ loadingConfigError: error, loadingConfig: false });
+    }
   }
 
   render() {
     const { extensions } = this.props;
+    const { loadingConfigError, loadingConfig } = this.state;
 
     const lang = messages[this.props.lang] ? this.props.lang : 'en';
     const logoutButton = (
       <LogoutButton getLogoutURL={() => this.props.logoutURL} />
     );
 
+    if (loadingConfig) {
```

Comment:
General comment / question before I review the rest in detail.

Do we need this global loading state? Is there something preventing us from beginning to render the application shell at least, and handling the loading state locally to the components that depend on the install properties?

Reply:
Not sure, maybe it could be structured differently but we need to load the config first before deciding if we have to fetch namespaces and adapt fetching extensions.

Reply:
It should be fine to get rid of this branch in the render, allow the app shell to start rendering, and we already respond to extension updates from the store.

`fetchConfig` also needs a small change to wait for the install properties to be loaded:

```diff
 async fetchConfig() {
     this.setState({ loadingConfig: true });
     try {
-      this.props.fetchInstallProperties();
+      await this.props.fetchInstallProperties();
       this.setState({ loadingConfig: false });
     } catch (error) {
       console.error(error); // eslint-disable-line no-console
       this.setState({ loadingConfigError: error, loadingConfig:
 false });
     }
   }
```

Since there's only a single namespace available to users when tenantNamespace is set we should also replace the dropdown in SideNav with text:
![image](https://user-images.githubusercontent.com/2829095/83017721-0dd7db00-a01c-11ea-8d54-9aab663bbd0e.png)

Reply:
Good catch with the missing `await`.
I will remove the `loadingConfig` branch in render to see how it goes.

I'm working on reorganizing `kustomize` bases/overlays right now, I will take care of this asap.

Reply:
If the app renders before the config is fetched, the user is redirected to the pipeline runs page and a call to load pipeline runs at the cluster level is initiated.

Of course this call fails in single namespace mode.

How should we handle this case ?

Reply:
I added the missing `await` and improved the sidenav to show the namespace when tenant namespace is set.

Reply:
@AlanGreene any advice on how to prevent blocking rendering until config is loaded ?

Reply:
hrmm interesting problem alright...

I'll take a look and see what options we have. I'd really like to avoid a full-page loading state if possible, or at least restrict it to only the routes affected.

Reply:
It looks like we already have this problem intermittently on the Secrets / ServiceAccounts pages (users gets redirected to PipelineRuns), and I'd expect to encounter it with extensions too but haven't managed to reproduce it there yet.

I've been trying out some changes to the Route components to see if we can handle the access checks more locally, building on what I already did with the ReadWriteRoute (which is working around some annoying react-router behaviour when dynamically adding routes) but so far it's not making things any simpler or really improving the experience.

We're going to need to handle more granular access control anyway, with each Route responsible for checking whether/how it should render its content. I have some thoughts that might be worth discussing / expanding upon at the WG but don't want to hold up the single namespace change (and other changes behind it...) until this is all resolved.

Maybe we can go with the full-page loading experience for now, just style it so it's not quite so bare, and do our best to improve it and minimise the instances where it's shown until we have a more complete / flexible solution in place.

Reply:
I will try to improve style a little bit. Let's see what i can do with css :P

Reply:
We could use the Carbon Loading component with the overlay, and see if we can add a string below the spinner 'Loading configuration...' or something.

The loading component should not cover the top navbar.

https://www.carbondesignsystem.com/patterns/loading-pattern#full-screen-loading
https://react.carbondesignsystem.com/?path=/story/loading--default

Reply:
I added the `Loading` component while fetching config.
There was no option to add text below the spinner though.

Reply:
Played around with this a bit and came up with something I think works well enough for now: https://github.com/tektoncd/dashboard/pull/1464

![image](https://user-images.githubusercontent.com/2829095/83701738-8cc3a980-a602-11ea-8953-fef89eef38d2.png)

```diff
diff --git a/src/containers/App/App.js b/src/containers/App/App.js
index b9a4349..627f123 100644
--- a/src/containers/App/App.js
+++ b/src/containers/App/App.js
@@ -22,10 +22,12 @@ import {
 } from 'react-router-dom';
 
 import { IntlProvider } from 'react-intl';
-import { Content, InlineNotification, Loading } from 'carbon-components-react';
+
+import { Content, InlineNotification } from 'carbon-components-react';
 
 import {
   Header,
+  LoadingShell,
   LogoutButton,
   PageErrorBoundary
 } from '@tektoncd/dashboard-components';
@@ -165,18 +167,6 @@ export /* istanbul ignore next */ class App extends Component {
       <LogoutButton getLogoutURL={() => this.props.logoutURL} />
     );
 
-    if (loadingConfig) {
-      return (
-        <IntlProvider
-          locale={lang}
-          defaultLocale="en"
-          messages={messages[lang]}
-        >
-          <Loading description="Loading configuration..." withOverlay />
-        </IntlProvider>
-      );
-    }
-
     return (
       <IntlProvider locale={lang} defaultLocale="en" messages={messages[lang]}>
         {loadingConfigError && (
@@ -188,6 +178,8 @@ export /* istanbul ignore next */ class App extends Component {
           />
         )}
 
+        {loadingConfig && <LoadingShell />}
+        {!loadingConfig && (
           <Router>
             <>
               <Header logoutButton={logoutButton} />
@@ -433,6 +425,7 @@ export /* istanbul ignore next */ class App extends Component {
               </Content>
             </>
           </Router>
+        )}
       </IntlProvider>
     );
   }
```

Let me know what you think.

Reply:
Wow that’s gold !!
Will try it this morning, thank you so much !

Reply:
Changed to use `LoadingShell`, works perfectly ! Thanks